staging/src/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go

```diff
@@ -1482,27 +1565,54 @@ func (az *Cloud) reconcilePublicIP(clusterName string, service *v1.Service, lbNa
 		return nil, err
 	}
 
-	var found bool
+	var serviceAnnotationRequestsNamedPublicIP bool = shouldPIPExisted
+	var discoveredDesiredPublicIP bool
+	var deletedDesiredPublicIP bool
 	var pipsToBeDeleted []*network.PublicIPAddress
 	for i := range pips {
 		pip := pips[i]
 		pipName := *pip.Name
-		if serviceOwnsPublicIP(&pip, clusterName, serviceName) {
-			// We need to process for pips belong to this service
-			if wantLb && !isInternal && pipName == desiredPipName {
-				// This is the only case we should preserve the
-				// Public ip resource with match service tag
-				found = true
-			} else {
-				pipsToBeDeleted = append(pipsToBeDeleted, &pip)
-			}
-		} else if wantLb && !isInternal && pipName == desiredPipName {
-			found = true
+		pipPropertiesFormat := pip.PublicIPAddressPropertiesFormat
+
+		// If we've been told to use a specific public ip by the client, let's track whether or not it actually existed
+		// when we inspect the set in Azure.
+		discoveredDesiredPublicIP = discoveredDesiredPublicIP || wantLb && !isInternal && pipName == desiredPipName
+
+		// Now, let's perform additional analysis.
+		// Specifically, we should decide if the discovered public ips should be released/deleted.
+
+		// If the service does not own the ip, we cannot delete it under any circumstances
+		if !serviceOwnsPublicIP(&pip, clusterName, serviceName) {
+			continue
+		}
+
+		if
+		// If we don't actually want a load balancer,
+		!wantLb ||
+			// If the load balancer is internal, and thus doesn't require public exposure
+			isInternal ||
+			// If the name of this public ip does not match the desired name,
+			(pipName != desiredPipName) ||
+			// If the service annotations have specified required ip tags, but thy do not match the iptags of the existing instance
+			(serviceIPTags != nil && (pipPropertiesFormat == nil /*unit test*/ ||
```

Comment:
I think we should use `len(serviceIPTags) != 0` here, as `make(map[string]string)` is not nil